Check architecture in upgrade_select module

chcao opened the pull request:
Add the architecture checking in upgrade_select module.

- Related ticket: https://progress.opensuse.org/issues/37946
- Verification run: 
Graphic interface:
https://openqa.nue.suse.com/tests/6164533#step/upgrade_select/2
https://openqa.nue.suse.com/tests/6164534#step/upgrade_select/2
https://openqa.nue.suse.com/tests/6164535#step/upgrade_select/2
https://openqa.nue.suse.com/tests/6190384#step/upgrade_select/2
https://openqa.nue.suse.com/tests/6164760#step/upgrade_select/2
Textmode:
https://openqa.nue.suse.com/tests/6164615#step/upgrade_select/2
https://openqa.nue.suse.com/tests/6164616#step/upgrade_select/2
https://openqa.nue.suse.com/tests/6164743#step/upgrade_select/2
https://openqa.nue.suse.com/tests/6191128#step/upgrade_select/2

tinawang123 commented:
Do you add needles for all platforms?

rfan1 commented on tests/installation/upgrade_select.pm:
```diff
@@ -39,7 +39,10 @@ sub run {
     if (match_has_tag("select-for-update")) {
-        send_key $cmd{next};
+        assert_screen('select-for-update-arch');
+        if (match_has_tag("select-for-update-arch")) {
```
Chenzi, 
You have more than one needles which have tag "select-for-update-arch",right?
if no, I suggest removing this line :)

chcao replied:
Yes, there are eight needles within this tag.

chcao commented:
> Do you add needles for all platforms?

Yes, I added the needles for four platforms.

rfan1 reviewed (approved):
LGTM

jouyingbin commented:
@chcao These VR are all about different platform with different Service packs , am I correct?

chcao commented:
> @chcao These VR are all about different platform with different Service packs , am I correct? Because I don't open each of them. :)

Yes, the VR includes offline media/pscc SLES,SLED, HA, HPC, from SLE12SPx and SLE15SPx, for four arches.

jouyingbin commented:
> > @chcao These VR are all about different platform with different Service packs , am I correct? Because I don't open each of them. :)
> 
> Yes, the VR includes offline media/pscc SLES,SLED, HA, HPC, from SLE12SPx and SLE15SPx, for four arches.

Thanks for information.

jouyingbin reviewed (approved):
LGTM.

coolgw commented on tests/installation/upgrade_select.pm:
```diff
@@ -39,7 +39,10 @@ sub run {
     }
     if (match_has_tag("select-for-update")) {
-        send_key $cmd{next};
+        assert_screen('select-for-update-arch');
```
I suggest using $arch as tag, $arch=setting(platform), then we check specific screen for specific platform.

chcao replied:
Thanks for comment, I changed the code as your suggestion.

lemon-suse commented on tests/installation/upgrade_select.pm:
```diff
@@ -39,7 +39,10 @@ sub run {
     if (match_has_tag("select-for-update")) {
-        send_key $cmd{next};
+        assert_screen('select-for-update-arch');
+        if (match_has_tag("select-for-update-arch")) {
```
This is not needed, since if the needle not matched then fail there.

lemon-suse commented on tests/installation/upgrade_select.pm:
```diff
@@ -39,7 +39,10 @@ sub run {
     }
     if (match_has_tag("select-for-update")) {
-        send_key $cmd{next};
+        assert_screen('select-for-update-arch');
```
Yes, here can be assert_screen("select-for-update-$arch"); to make it more specific for each platform.
$arch = get_var("ARCH");

chcao replied:
Thanks, I changed the code as your suggestion, please help to review.

lemon-suse commented on tests/installation/upgrade_select.pm:
```diff
@@ -39,6 +39,8 @@ sub run {
     }
     if (match_has_tag("select-for-update")) {
+        my $arch = get_var("ARCH");
+        assert_screen('select-for-update-' . "$arch");
```
I think you can write it like assert_screen("select-for-update-$arch");

chcao replied:
I have tried it, but it didn't work.

lemon-suse replied:
simple run on my test environment, no problem.

my $t = 'aaa';
my $a = "test-$t";
print "got a=$a";

got a=test-aaalemon@linux-hxjt:~/perl>

Please double check your code.

chcao replied:
I confirmed it doesn't work. When I see your comments above, I modified it as your suggestion assert_screen("select-for-update-$arch"), the verification case didn't pass, then I changed to this way. I don't want to try again.

chcao replied:
As discussed, I'll investigate why your method doesn't work, and won't change pr any more.

chcao replied:
Here is the test result:
I created a branch with your method: https://github.com/os-autoinst/os-autoinst-distri-opensuse/compare/master...chcao:test?expand=1
And here is test result: https://openqa.suse.de/tests/6220781#step/upgrade_select/2

lemon-suse replied:
- assert_screen('select-for-update-$arch');
Just as what I guess, you have make it wrong as ' ' ' while not ' " ', that why it not works.

lilyeyes reviewed (approved):
LGTM.

lilyeyes commented:
The way `assert_screen('select-for-update-' . "$arch");` is good too, I usually use this way: only create needed needles without code changes, e,g., select-for-update-1, select-for-update-2, select-for-update-3, ...

ggardet commented:
> > Do you add needles for all platforms?
> 
> Yes, I added the needles for four platforms.

No, there was no needle for aarch64. I needed to create it: https://github.com/os-autoinst/os-autoinst-needles-opensuse/commit/2b0826a378188384f831760e3ff579e2c443e88e

chcao commented:
> > > Do you add needles for all platforms?
> > 
> > 
> > Yes, I added the needles for four platforms.
> 
> No, there was no needle for aarch64. I needed to create it: [os-autoinst/os-autoinst-needles-opensuse@2b0826a](https://github.com/os-autoinst/os-autoinst-needles-opensuse/commit/2b0826a378188384f831760e3ff579e2c443e88e)

For aarch64, I added needles of  upgrade_select-select-for-update-aarch64-20210603.json and upgrade_select-select-for-update-aarch64-text-20210603.json.
Is your case for openSUSE product? I just verify the SLE products.

ggardet commented:
> Is your case for openSUSE product? I just verify the SLE products.

Yes, this is for openSUSE.

chcao commented:
> > > Do you add needles for all platforms?
> > 
> > 
> > Yes, I added the needles for four platforms.
> 
> No, there was no needle for aarch64. I needed to create it: [os-autoinst/os-autoinst-needles-opensuse@2b0826a](https://github.com/os-autoinst/os-autoinst-needles-opensuse/commit/2b0826a378188384f831760e3ff579e2c443e88e)

I checked the openSUSE cases, you have added the needles for the cases, thank you. If any more issues, please feel free to contact me, thanks.
